pwiz_tools/Skyline/Util/ParsedMolecule.cs

```diff
@@ -204,6 +206,90 @@ public static bool TryParseFormula(string formula, out ParsedMolecule resultMole
             }
         }
 
+        /// <summary>
+        /// Separate the formula and mass offset declarations in a string presumed to describe a chemical formula and/or mass offsets.
+        /// Is aware of simple math, e.g. "C12H5[-1.2/1.21]-C2H[-1.1]" => "C12H5-C2H", 0.1, 0.11
+        /// </summary>
+        /// <param name="formulaAndMasses">input string  e.g. "C12H5", "C12H5[-1.2/1.21]", "[+1.3/1.31]", "1.3/1.31", "C12H5[-1.2/1.21]-C2H[-1.1]"</param>
+        /// <param name="formula">string with any mass modifiers stripped out e.g.  "C12H5[-1.2/1.21]-C2H[-1.1]" ->  "C12H5-C2H" </param>
+        /// <param name="modMassMono">effect of any mono mass modifiers e.g. "C12H5[-1.2/1.21]-C2H[-1.1]" => 0.1 </param>
+        /// <param name="modMassAverage">effect of any avg mass modifiers e.g. "C12H5[-1.2/1.21]-C2H[-1.1]" => 0.11</param>
+        private static void SplitFormulaAndMasses(string formulaAndMasses, out string formula, out double? modMassMono, out double? modMassAverage)
+        {
+            modMassMono = null;
+            modMassAverage = null;
+            formula = formulaAndMasses?.Trim();
+            if (string.IsNullOrEmpty(formula))
+            {
+                return;
+            }
+            // A few different possibilities here, e.g. "C12H5", "C12H5[-1.2/1.21]", "[+1.3/1.31]", "1.3/1.31"
+            // Also possibly  "C12H5[-1.2/1.21]-C2H[-1.1]"
```

Comment:
Really feels like this code could benefit from some sub-functions, like something that just parses the inside of [<offset>], and something that parses <formula>[<offset>]. This looks like challenging code to maintain.

Reply:
Yeah. Just a little more like recursive descent parsing, since it is allowed to be:
{expression} = {formula}[{offset}]
or
{expression}-{expression}

I may have set a bad example elsewhere when I originally added {formula}-{formula} for peptide modifications. I am not sure of that. But, the addition of the multi-format offset:

{offset} = {double}
{offset} = {double}/{double}

Reaches the point where I would really like to see it in recursive descent and not a loop.

Reply:
I encapsulated the recursion in a regular expression. It reads pretty cleanly now, I think.